core/resp.go

```diff
@@ -2,134 +2,124 @@ package core
+func readLength(buf *bytes.Buffer) (int64, error) {
+	s, err := readStringUntilSr(buf)
+	if err != nil {
+		return 0, err
 	}
-	return 0, 0
-}
 
-// reads a RESP encoded simple string from data and returns
-// the string, the delta, and the error
-func readSimpleString(data []byte) (string, int, error) {
-	// first character +
-	pos := 1
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return 0, err
+	}
 
-	for ; data[pos] != '\r'; pos++ {
+	return v, nil
+}
+
+func readStringUntilSr(buf *bytes.Buffer) (string, error) {
+	s, err := buf.ReadString('\r')
+	if err != nil {
+		return "", err
 	}
+	// increamenting to skip `\n`
+	buf.ReadByte()
+	return s[:len(s)-1], nil
+}
 
-	return string(data[1:pos]), pos + 2, nil
+// reads a RESP encoded simple string from data and returns
+// the string and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readSimpleString(c io.ReadWriter, buf *bytes.Buffer) (string, error) {
+	return readStringUntilSr(buf)
 }
 
 // reads a RESP encoded error from data and returns
-// the error string, the delta, and the error
-func readError(data []byte) (string, int, error) {
-	return readSimpleString(data)
+// the error string and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readError(c io.ReadWriter, buf *bytes.Buffer) (string, error) {
+	return readStringUntilSr(buf)
 }
 
 // reads a RESP encoded integer from data and returns
-// the intger value, the delta, and the error
-func readInt64(data []byte) (int64, int, error) {
-	// first character :
-	pos := 1
-	var value int64 = 0
-
-	for ; data[pos] != '\r'; pos++ {
-		value = value*10 + int64(data[pos]-'0')
+// the intger value and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readInt64(c io.ReadWriter, buf *bytes.Buffer) (int64, error) {
+	s, err := readStringUntilSr(buf)
+	if err != nil {
+		return 0, err
+	}
+
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return 0, err
 	}
 
-	return value, pos + 2, nil
+	return v, nil
 }
 
 // reads a RESP encoded string from data and returns
-// the string, the delta, and the error
-func readBulkString(data []byte) (string, int, error) {
-	// first character $
-	pos := 1
+// the string, and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readBulkString(c io.ReadWriter, buf *bytes.Buffer) (string, error) {
+	len, err := readLength(buf)
+	if err != nil {
+		return "", err
+	}
 
-	// reading the length and forwarding the pos by
-	// the lenth of the integer + the first special character
-	len, delta := readLength(data[pos:])
-	pos += delta
+	var bytesRem int64 = len + 2 // 2 for \r\n
```

Comment:
Alignment issues on 32-bit machines, int64 would compile fine and it would give you full 64 bits wide integer but it also enforces alignment as it would take 2 cycles to access such an integer. This is part of micro-optimization.